src/main/java/com/bpodgursky/jbool_expressions/Not.java

```diff
@@ -14,6 +15,7 @@
 
   private Not(Expression<K> e) {
     this.e = e;
+    this.hash = 0 - e.hash;
```

Comment:
I think you designed this so that Not(Not(e)) and e would have the same hash code, but I actually find that a bit dubious since they aren't the same expression (how can you tell if you optimized it?)

Reply:
the same reasoning would go for (A | B) vs. (B | A). Are they the same expression? (spoiler: I have no satisfactory answer to this).

Reply:
In any case, `equals()` would return `false` for `Not(Not(A))` and `A`, as the initial types wouldn't match. In the case of more layers of `Not`, each layer flips from positive to negative, so unless you have the same number of `Not` layers on each expressions (making the expressions equal?) `equals()` will still return `false` (I discuss this in terms of `equals` as the bottom line is that the hashes are compared for equality).